lib/Parser/parse.cpp

```diff
@@ -7763,10 +7797,11 @@ PidRefStack* Parser::PushPidRef(IdentPtr pid)
     Assert(GetCurrentBlock() != nullptr);
     AssertMsg(pid != nullptr, "PID should be created");
     PidRefStack *ref = pid->GetTopRef();
-    if (!ref || (ref->GetScopeId() < GetCurrentBlock()->sxBlock.blockId
-                // We could have the ref from the parameter scope. In that case we can skip creating a new one.
+    if (!ref || (ref->GetScopeId() < GetCurrentBlock()->sxBlock.blockId)
+                // We could have the ref from the parameter scope if it is merged with body scope. In that case we can skip creating a new one.
                 && !(m_currentBlockInfo->pBlockInfoOuter->pnodeBlock->sxBlock.blockType == PnodeBlockType::Parameter
-                    && m_currentBlockInfo->pBlockInfoOuter->pnodeBlock->sxBlock.blockId == ref->GetScopeId())))
+                    && m_currentBlockInfo->pBlockInfoOuter->pnodeBlock->sxBlock.blockId == ref->GetScopeId()
+                    && (m_currentBlockInfo->pBlockInfoOuter->pnodeBlock->sxBlock.scope == nullptr || m_currentBlockInfo->pBlockInfoOuter->pnodeBlock->sxBlock.scope->GetCanMergeWithBodyScope())))
```

Comment:
nit: can you break this line - it is hard to understand if you have to horiz-scroll multiple times.